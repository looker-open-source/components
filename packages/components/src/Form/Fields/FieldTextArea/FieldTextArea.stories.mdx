import { FieldTextArea } from '@looker/components'

<Meta component={FieldTextArea} title="Docs/Components/Fields/FieldTextArea" />

# FieldTextArea

The `<FieldTextArea />` component is composed of `<TextArea />` and `<Label />` components. Using `<FieldTextArea />` allows for rendering validation messages.

```jsx
<FieldTextArea
  label="FieldTextArea Label"
  validationMessage={{ message: 'Some extra information' }}
/>
```
