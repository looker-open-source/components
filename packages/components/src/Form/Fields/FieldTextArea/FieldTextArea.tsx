/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { FC } from 'react'
import styled from 'styled-components'
import { useID } from '../../../utils'
import { useFormContext } from '../../Form'
import { TextArea, TextAreaProps } from '../../Inputs/TextArea'
import { Field, FieldProps, omitFieldProps, pickFieldProps } from '../Field'

export interface FieldTextAreaProps extends FieldProps, TextAreaProps {}

const FieldTextAreaComponent: FC<FieldTextAreaProps> = ({ ...props }) => {
  const id = useID(props.id)
  const validationMessage = useFormContext(props)
  return (
    <Field
      {...pickFieldProps(props)}
      id={id}
      validationMessage={validationMessage}
    >
      <TextArea
        {...omitFieldProps(props)}
        aria-describedby={`${id}-describedby`}
        validationType={validationMessage && validationMessage.type}
      />
    </Field>
  )
}

FieldTextAreaComponent.displayName = 'FieldTextAreaComponent'

export const FieldTextArea = styled(FieldTextAreaComponent)``
