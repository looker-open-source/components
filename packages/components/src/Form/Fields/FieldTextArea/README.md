# FieldTextArea

The `<FieldTextArea />` component is composed of `<TextArea />` and `<Label />` components. Using `<FieldTextArea />` allows for rendering validation messages.

```tsx
function Basic(props: FieldTextAreaProps) {
  const { name = 'firstName', label = 'First Name', ...restProps } = props;
  return <FieldTextArea name={name} label={label} {...restProps} />;
}
```

## Controlled

```tsx
function Controlled(props: FieldTextAreaProps) {
  const {
    name: _name,
    label: _label,
    value: valueProp = 'Initial Value',
    ...restProps
  } = props;
  const [value, setValue] = useState(valueProp);
  const handleReset = () => setValue(valueProp);
  const handleClear = () => setValue('');
  const handleChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
    setValue(e.target.value);
  };
  return (
    <SpaceVertical>
      <Space>
        <Button onClick={handleReset}>Reset</Button>
        <Button onClick={handleClear}>Clear</Button>
      </Space>
      <FieldTextArea
        width="100%"
        label="Controlled"
        value={value}
        onChange={handleChange}
        {...restProps}
      />
    </SpaceVertical>
  );
}
```

## Default Value

```tsx
function DefaultValue(props: FieldTextAreaProps) {
  const {
    name = 'firstName',
    label = 'First Name',
    defaultValue = 'Default value',
    ...restProps
  } = props;
  return (
    <FieldTextArea
      name={name}
      label={label}
      defaultValue={defaultValue}
      {...restProps}
    />
  );
}
```

## Detail & Description

```tsx
function DetailDescription(props: FieldTextAreaProps) {
  const {
    name = 'firstName',
    label = 'First Name',
    description = 'This is a description',
    detail = '0/50',
    ...restProps
  } = props;

  return (
    <FieldTextArea
      name={name}
      label={label}
      description={description}
      detail={detail}
      {...restProps}
    />
  );
}
```

## Disabled

```tsx
function DisabledStory(props: FieldTextAreaProps) {
  const {
    name = 'firstName',
    label = 'First Name',
    defaultValue = 'Default value',
    disabled = true,
    ...restProps
  } = props;

  return (
    <FieldTextArea
      name={name}
      label={label}
      defaultValue={defaultValue}
      disabled={disabled}
      {...restProps}
    />
  );
}
```

## Floating Label

```tsx
function FloatingLabel(props: FieldTextAreaProps) {
  const {
    name = 'firstName',
    label = 'First Name',
    description = 'Some important information about this field',
    detail = '0/50',
    ...restProps
  } = props;
  return (
    <ExtendComponentsThemeProvider
      themeCustomizations={{ defaults: { externalLabel: false } }}
    >
      <FieldTextArea
        detail={detail}
        description={description}
        name={name}
        label={label}
        {...restProps}
      />
    </ExtendComponentsThemeProvider>
  );
}
```

### Floating Label with Default Value

```tsx
function FloatingLabelDefaultValue(props: FieldTextAreaProps) {
  const {
    name = 'firstName',
    label = 'First Name',
    description = 'Some important information about this field',
    detail = '0/50',
    defaultValue = 'My Name',
    ...restProps
  } = props;
  return (
    <ExtendComponentsThemeProvider
      themeCustomizations={{ defaults: { externalLabel: false } }}
    >
      <FieldTextArea
        detail={detail}
        description={description}
        name={name}
        label={label}
        defaultValue={defaultValue}
        {...restProps}
      />
    </ExtendComponentsThemeProvider>
  );
}
```

## Inline

```tsx
function Inline(props: FieldTextAreaProps) {
  const {
    inline = true,
    name = 'firstName',
    label = 'First Name',
    detail = 'Your preferred salutation',
    ...restProps
  } = props;

  return (
    <FieldTextArea
      name={name}
      label={label}
      inline={inline}
      detail={detail}
      {...restProps}
    />
  );
}
```

## Disabling Textare Resizing

```tsx
function NoResize(props: FieldTextAreaProps) {
  const {
    resize = false,
    name = 'firstName',
    label = 'First Name',
    ...restProps
  } = props;

  return (
    <FieldTextArea name={name} label={label} resize={resize} {...restProps} />
  );
}
```

## Required

```tsx
function Required(props: FieldTextAreaProps) {
  const {
    required = true,
    name = 'firstName',
    label = 'First Name',
    ...restProps
  } = props;

  return (
    <FieldTextArea
      name={name}
      label={label}
      required={required}
      {...restProps}
    />
  );
}
```

## Validation Message

```tsx
function Validation(props: FieldTextAreaProps) {
  const {
    name = 'firstName',
    label = 'First Name',
    validationMessage = { message: 'Some extra information', type: 'error' },
    ...restProps
  } = props;

  return (
    <FieldTextArea
      name={name}
      label={label}
      validationMessage={validationMessage}
      {...restProps}
    />
  );
}
```
