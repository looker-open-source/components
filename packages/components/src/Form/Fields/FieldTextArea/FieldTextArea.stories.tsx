/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { Story } from '@storybook/react/types-6-0'
import type { FormEvent } from 'react'
import React, { useState } from 'react'
import { ExtendComponentsThemeProvider } from '@looker/components-providers'
import { defaultArgTypes as argTypes } from '../../../../../../apps/storybook/src/defaultArgTypes'
import { Button } from '../../../Button'
import { Space, SpaceVertical } from '../../../Layout'
import type { FieldTextAreaProps } from './FieldTextArea'
import { FieldTextArea } from './FieldTextArea'

export default {
  argTypes,
  component: FieldTextArea,
  title: 'FieldTextArea',
}

const Template: Story<FieldTextAreaProps & { externalLabel: boolean }> = ({
  externalLabel = true,
  ...args
}) => (
  <ExtendComponentsThemeProvider
    themeCustomizations={{ defaults: { externalLabel } }}
  >
    <FieldTextArea {...args} />
  </ExtendComponentsThemeProvider>
)

export const Basic = Template.bind({})
Basic.args = { label: 'Text Area' }

export const FloatingLabel = Template.bind({})
FloatingLabel.args = { ...Basic.args, externalLabel: false }

export const FloatingLabelValue = Template.bind({})
FloatingLabelValue.args = {
  ...Basic.args,
  defaultValue: 'Default value',
  externalLabel: false,
}

export const DefaultValue = Template.bind({})
DefaultValue.args = { ...Basic.args, defaultValue: 'Default value' }

export const Disabled = Template.bind({})
Disabled.args = { ...Basic.args, disabled: true }

export const DisabledValue = Template.bind({})
DisabledValue.args = { ...DefaultValue.args, disabled: true }

export const Required = Template.bind({})
Required.args = { ...Basic.args, required: true }

export const Description = Template.bind({})
Description.args = { ...Basic.args, description: 'This is a description' }

export const Detail = Template.bind({})
Detail.args = { ...Basic.args, detail: '0/50' }

export const DetailDescription = Template.bind({})
DetailDescription.args = {
  ...Basic.args,
  description: 'This is a description',
  detail: '0/50',
}

export const Error = Template.bind({})
Error.args = {
  ...Basic.args,
  validationMessage: { message: 'Error Message', type: 'error' },
}

export const ErrorDetail = Template.bind({})
ErrorDetail.args = {
  ...Basic.args,
  detail: '0/50',
  validationMessage: { message: 'Error Message', type: 'error' },
}

export const ErrorValueDetail = Template.bind({})
ErrorValueDetail.args = {
  ...Basic.args,
  defaultValue: 'This value is too long',
  detail: '50/50',
  validationMessage: { message: 'Error Message', type: 'error' },
}

export const Inline = Template.bind({})
Inline.args = {
  ...Basic.args,
  inline: true,
}

export const NoResize = Template.bind({})
NoResize.args = {
  placeholder: 'no resize',
  resize: false,
}

export const ExternalLabel = Template.bind({})
ExternalLabel.args = {
  ...DetailDescription.args,
  externalLabel: true,
}

const initialValue = 'Initial Value'

export const Controlled = () => {
  const [value, setValue] = useState(initialValue)
  const handleReset = () => setValue(initialValue)
  const handleClear = () => setValue('')
  const handleChange = (e: FormEvent<HTMLTextAreaElement>) => {
    setValue(e.currentTarget.value)
  }
  return (
    <SpaceVertical>
      <Space>
        <Button onClick={handleReset}>Reset</Button>
        <Button onClick={handleClear}>Clear</Button>
      </Space>
      <FieldTextArea
        width="100%"
        label="Controlled"
        value={value}
        onChange={handleChange}
      />
    </SpaceVertical>
  )
}

Controlled.parameters = {
  storyshots: { disable: true },
}
