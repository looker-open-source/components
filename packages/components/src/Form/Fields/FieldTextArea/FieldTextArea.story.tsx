/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { Story } from '@storybook/react/types-6-0'
import React from 'react'
import { FieldTextArea, FieldTextAreaProps } from './FieldTextArea'

export default {
  component: FieldTextArea,
  title: 'FieldTextArea',
}

const Template: Story<FieldTextAreaProps> = (args) => (
  <FieldTextArea {...args} />
)

export const Basic = Template.bind({})
Basic.args = { label: 'Text Area' }

export const Disabled = Template.bind({})
Disabled.args = { ...Basic.args, disabled: true }

export const Required = Template.bind({})
Required.args = { ...Basic.args, required: true }

export const Error = Template.bind({})
Error.args = {
  ...Basic.args,
  validationMessage: { message: 'Error Message', type: 'error' },
}

export const Inline = Template.bind({})
Inline.args = {
  ...Basic.args,
  inline: true,
}

export const NoResize = Template.bind({})
NoResize.args = {
  placeholder: 'no resize',
  resize: false,
}
