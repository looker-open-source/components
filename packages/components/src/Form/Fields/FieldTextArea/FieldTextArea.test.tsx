/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import React from 'react'
import {
  assertSnapshot,
  mountWithTheme,
  renderWithTheme,
} from '@looker/components-test-utils'
import { FieldTextArea } from './FieldTextArea'

test('A FieldTextArea with default label', () => {
  assertSnapshot(<FieldTextArea id="FieldTextAreaID" label="👍" />)
})

test('A FieldTextArea with label inline', () => {
  assertSnapshot(<FieldTextArea id="FieldTextAreaID" inline label="👍" />)
})

test('A FieldTextArea required', () => {
  const wrapper = mountWithTheme(
    <FieldTextArea id="FieldTextAreaID" label="👍" required />
  )

  expect(wrapper.text()).toMatch(`👍 required`)
})

test('A FieldTextArea disabled', () => {
  const wrapper = mountWithTheme(
    <FieldTextArea disabled id="FieldTextAreaID" label="👍" />
  )
  wrapper.find('textarea').html().includes('disabled=""')
})

test('A FieldTextArea with description has proper aria setup', () => {
  const description = 'This is a description'

  const { container, getByDisplayValue } = renderWithTheme(
    <FieldTextArea id="test" defaultValue="example" description={description} />
  )

  const input = getByDisplayValue('example')
  const id = input.getAttribute('aria-describedby')
  expect(id).toBeDefined()

  const describedBy = container.querySelector(`#${id}`)
  expect(describedBy).toHaveTextContent(description)
})

test('A FieldTextArea with error has proper aria setup', () => {
  const errorMessage = 'This is an error'

  const { container, getByDisplayValue } = renderWithTheme(
    <FieldTextArea
      id="test"
      defaultValue="example"
      validationMessage={{ message: errorMessage, type: 'error' }}
    />
  )

  const input = getByDisplayValue('example')
  const id = input.getAttribute('aria-describedby')
  expect(id).toBeDefined()

  const describedBy = container.querySelector(`#${id}`)
  expect(describedBy).toHaveTextContent(errorMessage)
})

test('A FieldTextArea with detail', () => {
  const wrapper = mountWithTheme(
    <FieldTextArea
      detail="5/50"
      id="FieldTextAreaID"
      label="hello"
      placeholder="placeholder"
    />
  )
  expect(wrapper.text()).toMatch(`hello5/50`)
})

test('A FieldTextArea with validationMessage', () => {
  const wrapper = mountWithTheme(
    <FieldTextArea
      id="FieldTextAreaID"
      label="hello"
      validationMessage={{
        message: 'validation Message',
        type: 'error',
      }}
      placeholder="placeholder"
    />
  )
  expect(wrapper.text()).toMatch(`helloWarningvalidation Message`)
})

test('FieldTextArea supports onChange handler', () => {
  let counter = 0
  const handleChange = () => counter++

  const wrapper = mountWithTheme(
    <FieldTextArea id="FieldTextAreaID" onChange={handleChange} />
  )

  wrapper.find('textarea').simulate('change', { target: { value: '' } })
  expect(counter).toEqual(1)
})
