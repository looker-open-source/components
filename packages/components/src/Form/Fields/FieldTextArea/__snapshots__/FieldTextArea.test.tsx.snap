// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A FieldTextArea 1`] = `
.c2 {
  width: 13rem;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  height: 6.25rem;
  width: 16rem;
  resize: both;
  border: solid 1px;
  border-color: #DEE1E5;
  border-radius: 0.25rem;
  font-size: 0.875rem;
  min-height: 6.25rem;
  padding: 0.75rem;
  padding-right: 2rem;
}

.c5:hover {
  border-color: #C1C6CC;
}

.c5:focus,
.c5:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c4 {
  position: relative;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

.c4 .c6 {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
}

.c0 {
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  color: #343C42;
  margin-right: 0.5rem;
  font-weight: 600;
  font-size: 0.75rem;
  display: inline-block;
}

.c1 {
  width: 13rem;
}

.c1 .c7 {
  width: 100%;
}

.c1 .c8 {
  width: 100%;
}

<div
  className="c0 c1 "
>
  <div
    className="c2"
    width="13rem"
  >
    <label
      className="c3"
      color="palette.charcoal700"
      fontSize="xsmall"
      fontWeight="semiBold"
      htmlFor="FieldTextArea"
    >
      this is a label
    </label>
  </div>
  <div
    className="c0"
  >
    <div
      className=""
    >
      <div
        className="c4 "
      >
        <textarea
          className="c5"
          height="6.25rem"
          width="16rem"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`A FieldTextArea with a validation message 1`] = `
.c2 {
  width: 13rem;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c6 {
  color: #CC1F36;
  width: 1rem;
  height: 1rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c5 {
  height: 6.25rem;
  width: 16rem;
  resize: both;
  border: solid 1px;
  border-color: #DEE1E5;
  border-radius: 0.25rem;
  font-size: 0.875rem;
  min-height: 6.25rem;
  padding: 0.75rem;
  padding-right: 2rem;
}

.c5:hover {
  border-color: #C1C6CC;
}

.c5:focus,
.c5:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c4 {
  position: relative;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

.c4 .c7 {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
}

.c0 {
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  color: #343C42;
  margin-right: 0.5rem;
  font-weight: 600;
  font-size: 0.75rem;
  display: inline-block;
}

.c8 {
  color: #CC1F36;
  margin-right: 0.5rem;
  margin-top: 0.5rem;
  font-size: 0.75rem;
}

.c1 {
  width: 13rem;
}

.c1 .c9 {
  width: 100%;
}

.c1 .c10 {
  width: 100%;
}

<div
  className="c0 c1 "
>
  <div
    className="c2"
    width="13rem"
  >
    <label
      className="c3"
      color="palette.charcoal700"
      fontSize="xsmall"
      fontWeight="semiBold"
      htmlFor="errorMessage"
    >
      this is a label
    </label>
  </div>
  <div
    className="c0"
  >
    <div
      className=""
    >
      <div
        className="c4 "
      >
        <textarea
          className="c5"
          height="6.25rem"
          width="16rem"
        />
        <div
          className="c6 c7 "
          color="palette.red500"
          size="1rem"
        >
          <svg
            fill="currentColor"
            height="100%"
            viewBox="0 0 22 19"
            width="100%"
          >
            <title>
              Warning
            </title>
            <path
              d="M0 19h22L11 0 0 19zm12-3h-2v-2h2v2zm0-4h-2V8h2v4z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="c8"
      fontSize="xsmall"
      type="error"
    >
      This is an error
    </div>
  </div>
</div>
`;
